extensions/amp-ad/0.1/amp-ad-xorigin-iframe-handler.js

```diff
@@ -161,37 +184,49 @@ export class AmpAdXOriginIframeHandler {
       this.element_.appendChild(this.iframe);
       return Promise.resolve();
     }
-    // Set iframe initially hidden which will be removed on load event +
-    // post message.
+
+    // Set iframe initially hidden which will be removed on render-start or
+    // load, whichever is earlier.
     setStyle(this.iframe, 'visibility', 'hidden');
     this.element_.appendChild(this.iframe);
+    Promise.race([
+      renderStartPromise,
+      iframeLoadPromise,
+      timer.promise(VISIBILITY_TIMEOUT),
+    ]).then(() => {
+      if (this.iframe) {
+        setStyle(this.iframe, 'visibility', '');
+        if (this.baseInstance_.emitLifecycleEvent) {
+          this.baseInstance_.emitLifecycleEvent('adSlotUnhidden');
+        }
+      }
+    });
 
-    return timerFor(this.baseInstance_.win).timeoutPromise(TIMEOUT_VALUE,
-        this.adResponsePromise_,
-        'timeout waiting for ad response').catch(e => {
-          this.noContent_();
-          user().warn('AMP-AD', e);
-        }).then(() => {
-          if (this.iframe) {
-            setStyle(this.iframe, 'visibility', '');
-            if (this.baseInstance_.emitLifecycleEvent) {
-              this.baseInstance_.emitLifecycleEvent('adSlotUnhidden');
-            }
-          }
-        });
+    // The actual ad load is eariliest of iframe.onload event and no-content.
+    const adLoadPromise = Promise.race([
+      iframeLoadPromise,
+      noContentPromise,
+    ]);
+    return adLoadPromise.catch(reason => {
```

Comment:
Removed completely. Not possible. And already well processed by the framework itself.